docs(docs-infra): Improve support of decorators in the API pages

JeanMeche opened the pull request:
fixes #60244

JeanMeche commented:
blocked on #60409

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/entities.mts:
```diff
@@ -16,6 +16,7 @@ export enum EntryType {
   Component = 'component',
   Constant = 'constant',
   Decorator = 'decorator',
+  InterfaceLikeDecorator = 'interface_like_decorator',
```
Question: Not entirely sure I understand the naming. Aren't all decorators have/described as interfaces?

devversion replied:
I think we should just name this `decorator`

JeanMeche replied:
We already has decorator though. 
My idea is to have 2 distinct code path for the 2 kinds of decorators.

devversion replied:
I'd need to look more at this PR, but I'd expect both types of decorators to share the same "entry type"

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.mts:
```diff
@@ -43,17 +43,36 @@ export function codeToHtml(code: string, language: string | undefined): string {
+}
+
+export function insertParenthesesForDecoratorInShikiHtml(shikiHtml: string): string {
+  // TODO: Look if we can use JSDom instead of regex
```
Yeah, I would support something like this since such complex regexes are both hard to read and has the possibility to fail due to `.*` matching everything.

JeanMeche replied:
I tried to add `jsdom` in the dependencies of `adev/shared-docs/pipeline/api-gen/rendering` but hit some issues with the bazel tooling.

hawkgs replied:
Yeah, it's just a comment that I agree it's something worth investing in the near future. I'll resolve this.

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/transforms/code-transforms.mts:
```diff
@@ -71,22 +80,33 @@ export async function addRenderableCodeToc<T extends DocEntry & HasModuleName>(
   }
 
+  // Note: Don't expect enum value in signatures to be linked correctly
+  // as skihi already splits them into separate span blocks.
```
Typo: skihi

hawkgs commented on packages/compiler-cli/src/ngtsc/docs/src/decorator_extractor.ts:
```diff
@@ -65,46 +82,105 @@ function getDecoratorType(declaration: ts.VariableDeclaration): DecoratorType |
+}
+
+/**
+ * @returns Interface properties for decorators that are akin to interfaces eg. @Component
+ * @returns null for decorators that are akin to functions eg. @Inject
```
Has two JSDoc returns.

hawkgs commented on packages/compiler-cli/src/ngtsc/docs/src/decorator_extractor.ts:
```diff
@@ -117,28 +193,86 @@ function getDecoratorOptions(
+  const replace: {initial: string; replacedWith: string}[] = [];
+  if (type.isUnion()) {
+    // The parameter can be a union, this includes optional parameters whiceh are a union of the type and undefined.
+    type.types.forEach((subType) => {
```
Style: `for ... of`?

hawkgs commented:
Since I am not very familiar with this part of the docs-infra and I am not sure that I can give a meaningful feedback, I'll let other reviewers chime in as well.

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/templates/decorator-reference.tsx:
```diff
@@ -0,0 +1,27 @@
+import {API_REFERENCE_CONTAINER} from '../styling/css-classes.mjs';
+import {DecoratorEntryRenderable} from '../entities/renderables.mjs';
+
+/** Component to render a class API reference document. */
```
The comment is not matching the new component.

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/templates/decorator-reference.tsx:
```diff
@@ -0,0 +1,27 @@
+/** Component to render a class API reference document. */
+export function DecoratorReference(entry: DecoratorEntryRenderable) {
+  return (
+    <div className={API_REFERENCE_CONTAINER}>
+      <HeaderApi entry={entry} />
+      <SectionApi entry={entry} />
+      <SectionDescription entry={entry} />
+      <SectionUsageNotes entry={entry} />
+    </div>
```
Question: The reference member cards won't exist for these decorators?

JeanMeche replied:
Yes, we should be able to match the same behavior as functions. 
Let me see what I can do

josephperrott commented on adev/shared-docs/pipeline/api-gen/extraction/index.mts:
```diff
@@ -125,7 +125,7 @@ function getEntriesFromMembers(entry: DocEntry): string[] {
 
 function hasMembers(entry: DocEntry): entry is InterfaceEntry | ClassEntry {
-  return 'members' in entry;
+  return 'members' in entry && Array.isArray(entry.members);
```
I believe you could just drop the assertion that `members` is in the `entry` as if its not it will return `undefined` and then `isArray` will be false .

```suggestion
  return Array.isArray(entry.members);
```

JeanMeche replied:
`members` is not present in the base class `DocEntry` but is in Interface/Class entries

josephperrott reviewed (approved):
LGTM

JeanMeche committed:
docs: improve support for decorator APIs

The commit introduce the distinction between

- class like decorators (like Component, Interface, NgModule. They are formatted like classes/interfaces, with each attribute being documents.
- function like decorators (Attribute, Host, Optional...)

devversion commented on adev/shared-docs/pipeline/api-gen/rendering/shiki/shiki.mts:
```diff
@@ -43,17 +43,36 @@ export function codeToHtml(code: string, language: string | undefined): string {
+  );
+}
+
+export function insertParenthesesForDecoratorInShikiHtml(shikiHtml: string): string {
```
I think a comment on why this is needed would be good.

JeanMeche replied:
I agree with you and listening to any suggestion. 

One todo I have inmind once we've migrated to toolchain is to drop all the regexes an use `jsdom` for dom manipulations.

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/type_extractor.ts:
```diff
@@ -10,9 +10,15 @@ import ts from 'typescript';
   );
+
+  // Remove the `undefined` union from the type string if the property is optional
+  if ((ts.isPropertyDeclaration(node) || ts.isPropertySignature(node)) && node.questionToken) {
```
Should this be configurable in `extractResolvedTypeString`? I think the function shouldn't magically do this.

devversion reviewed (approved):
LGTM, but this docs pipeline code is getting more and more complicated. Do you agree? I wonder if we can make some changes to improve readability for the long term?

thePunderWoman commented:
This PR was merged into the repository by commit 5c54cb87a5ea52e5b5e80f30181564de63e69e7a.

The changes were merged into the following branches: main, 20.1.x
